offline_pitch_extractor/yin_extractor.py

```diff
@@ -0,0 +1,230 @@
+import numpy as np
+import matplotlib.pyplot as plt
+from numpy import long
+from scipy.signal import fftconvolve
+from os import sep
+from audio_processing import *
+import time
+
+
+def differenceFunction_original(x, N, tau_max):
+    """
+    Compute difference function of data x. This corresponds to equation (6) in [1]
+    Original algorithm.
+    :param x: audio data
+    :param N: length of data
+    :param tau_max: integration window size
+    :return: difference function
+    :rtype: list
+    """
+    df = [0] * tau_max
+    for tau in range(1, tau_max):
+         for j in range(0, N - tau_max):
+             tmp = long(x[j] - x[j + tau])
+             df[tau] += tmp * tmp
+    return df
+
+
+def differenceFunction_scipy(x, N, tau_max):
+    """
+    Compute difference function of data x. This corresponds to equation (6) in [1]
+    Faster implementation of the difference function.
+    The required calculation can be easily evaluated by Autocorrelation function or similarly by convolution.
+    Wiener–Khinchin theorem allows computing the autocorrelation with two Fast Fourier transforms (FFT), with time complexity O(n log n).
+    This function use an accellerated convolution function fftconvolve from Scipy package.
+    :param x: audio data
+    :param N: length of data
+    :param tau_max: integration window size
+    :return: difference function
+    :rtype: list
+    """
+    x = np.array(x, np.float64)
+    w = x.size
+    x_cumsum = np.concatenate((np.array([0]), (x * x).cumsum()))
+    conv = fftconvolve(x, x[::-1])
+    tmp = x_cumsum[w:0:-1] + x_cumsum[w] - x_cumsum[:w] - 2 * conv[w - 1:]
+    return tmp[:tau_max + 1]
+
+
+
+
+def differenceFunction(x, N, tau_max):
+    """
+    Compute difference function of data x. This corresponds to equation (6) in [1]
+    Fastest implementation. Use the same approach than differenceFunction_scipy.
+    This solution is implemented directly with Numpy fft.
+    :param x: audio data
+    :param N: length of data
+    :param tau_max: integration window size
+    :return: difference function
+    :rtype: list
+    """
+
+    x = np.array(x, np.float64)
+    w = x.size
+    tau_max = min(tau_max, w)
+    x_cumsum = np.concatenate((np.array([0.]), (x * x).cumsum()))
+    size = w + tau_max
+    p2 = (size // 32).bit_length()
+    nice_numbers = (16, 18, 20, 24, 25, 27, 30, 32)
+    size_pad = min(x * 2 ** p2 for x in nice_numbers if x * 2 ** p2 >= size)
+    fc = np.fft.rfft(x, size_pad)
+    conv = np.fft.irfft(fc * fc.conjugate())[:tau_max]
+    return x_cumsum[w:w - tau_max:-1] + x_cumsum[w] - x_cumsum[:tau_max] - 2 * conv
```

Comment:
Maybe we should use easy words to describe the variable since it is easier to read and understand.

Reply:
FACTS